jvb/src/main/kotlin/org/jitsi/videobridge/Endpoint.kt

```diff
@@ -1239,4 +1359,379 @@ class Endpoint @JvmOverloads constructor(
 
         override fun trace(f: () -> Unit) = f.invoke()
     }
+
+    /**
+     * Align common fields from different source types.
+     * Perhaps this could become a base class of those types.
+     */
+    private class SourceDesc private constructor(
+        val name: String,
+        val owner: String,
+        val videoType: VideoType,
+        val ssrc1: Long,
+        val ssrc2: Long
+    ) {
+        constructor(s: AudioSourceDesc) : this(
+            s.sourceName ?: "anon", s.owner ?: "unknown", VideoType.DISABLED, s.ssrc, -1
+        )
+        constructor(s: MediaSourceDesc) : this(
+            s.sourceName ?: "anon", s.owner ?: "unknown", s.videoType, s.primarySSRC, getRtx(s)
+        )
+        companion object {
+            fun getRtx(s: MediaSourceDesc): Long {
+                /* Ignoring any additional entries for now. */
+                return s.rtpEncodings[0].getSecondarySsrc(SsrcAssociationType.RTX)
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        override fun toString(): String = "$owner:$ssrc1/$ssrc2"
+    }
+
+    /**
+     * Some RTP header state to track.
+     */
+    private class RtpState() {
+        var lastSequenceNumber = 0
+        var lastTimestamp = 0L
+        var lastTl0Index = -1
+        var valid = false
+
+        fun update(packet: RtpPacket) {
+            lastSequenceNumber = packet.sequenceNumber
+            lastTimestamp = packet.timestamp
+            lastTl0Index = if (packet is Vp9Packet) packet.TL0PICIDX else -1
+            valid = true
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        override fun toString(): String = if (valid) "$lastSequenceNumber/$lastTimestamp/$lastTl0Index" else "-"
+    }
+
+    /**
+     * RTP state for received SSRCs.
+     */
+    private class ReceiveSsrc(val props: SourceDesc) {
+        val state = RtpState()
+
+        /**
+         * True when sequence number and timestamp deltas have been calculated.
+         * If false, calculate them on the next relayed packet.
+         */
+        var hasDeltas = false
+
+        /**
+         * {@inheritDoc}
+         */
+        override fun toString(): String = "$state" + if (hasDeltas) "" else " (no \u2206)"
+    }
+
+    /**
+     * RTP state for sent SSRCs.
+     */
+    private class SendSsrc(ssrc_: Long?) {
+        val ssrc = ssrc_ ?: getNextSsrc()
+        private val state = RtpState()
+        private var sequenceNumberDelta = 0
+        private var timestampDelta = 0L
+        private var tl0IndexDelta = 0
+
+        companion object {
+            private val useRandom = true; /* switch off to ease debugging */
+            private var nextSsrc = 777000001L /* use serial number for debugging */
+            private val random = Random(System.currentTimeMillis())
```

Comment:
It might be better if this were cryptographically random - if not, you may want to use ThreadLocalRandom.

You may also want to check to avoid colliding with existing SSRCs?  I'm not sure whether you need to worry about that.

Reply:
Added checks for collisions. 
Switched to ThreadLocalRandom.
Can discuss further whether to use a better random.

Reply:
Switched to SecureRandom.